### PR TITLE
Integration test

### PR DESCRIPTION
**Pull request by xfred81**

### CODE REVIEW

**Comment by LudovicRousseau** [comment]
Please rebase your code above grisbi/master to have a minimal only patch.

**Comment by LudovicRousseau** [comment]
It is also recommended to use a git graphical client, like https://desktop.github.com/, to check your patches before you push and do a Pull Request.

**Comment by xfred81** [comment]
I though my various tests had no effects on other repositories. Sorry for that. My big challenge is that I try working on Windows port without having Windows. For various reasons, I mainly have to work with github's site for /editing/ files (which is not always simple).

BTW, it's my first git based project and I think I will check git's user guide about rebasing/merging/pulling etc. I'm far too unconformable with this tool.

Sorry again for inconveniences (btw, I managed to make a new Windows build that should include all latest changes you guys introduced past weeks).

**Comment by LudovicRousseau** [comment]
Exact. Your tests have no effect on other repositories. Also your work is really appreciated.

Maybe you can work on your own fork https://github.com/xfred81/grisbi and create Pull Requests only when your patches are ready?
I also highly suggest to work in a "test" (or whatever you want to call it) branch and not in the "master" branch. I will be easier to include/merge changes from the upstream grisbi repository.

Git can be complex to learn. Do not hesitate to ask for support.

**Comment by xfred81** [comment]
I did a "test" branch (grisbi_patch_for_win_targets_ok), which I - hopefull - use. Well, I didn't mean to use master, btw.

I confess I much more keen with CVS/SVN. Terms with git are somewhat different, and, as the "rebase" stuff, looks like quite specific concepts. Any advice about a decent tutorial about git?

**Comment by LudovicRousseau** [comment]
I do not know a decent git tutorial. You can find plenty of them on the web.

I notice that your repository is "disconnected" from the official repo. You can see that at https://github.com/xfred81/grisbi/network

You should "resync" your repository. I guess you have a "upstream" source as documented in https://help.github.com/articles/fork-a-repo/
I suggest you to do:
```
git checkout master
git rebase upstream/master
```
This will bring your master branch in sync with the official master branch. The 2 branches master and upstream/master should point to the same patch (unless you have local patches).

Then apply your own patches above master
```
git checkout grisbi_patch_for_win_targets_ok
git rebase master
```

You really should use a graphical tool to examine the git history. On GNU/Linux I gitk or gitg. On macOS I use GitX.

**Comment by xfred81** [comment]
Hm tried to apply those on a brand, new "clone" on my Linux box; yet the "git rebase upstream/master" command complains (_Needed a single revision / invalid upstream upstream_) . Tried "git rebase master" which logically returns _Current branch HEAD is up to date._.
The "upstream" aspect has been neglected here, I suppose that might be the point (?)

To be honest I feel like I just miss some little keys here, and I don't really have spare time to get depth knowledge on git (yep, holidays are over). I sincerely hope I'm not just boring you to much with such threads so... Goal is just to quickly adopt a more classic usage of git so that I won't disturb you with tons of unwanted commits or such.

**Comment by LudovicRousseau** [comment]
Do: ``git remote -v``. If you do not have any "upstream" line then do:
```
git remote add upstream https://github.com/grisbi/grisbi.git
git fetch upstream
git rebase upstream/master
```

**Comment by xfred81** [comment]
Great, a step beyond; I did this. Got auto-merging issues that gave me github while I tried to handle this through web interface. Errors are:
_Auto-merging src/ui/prefs_page_divers.ui
CONFLICT (content): Merge conflict in src/ui/prefs_page_divers.ui
Auto-merging src/prefs/prefs_page_divers.c
CONFLICT (content): Merge conflict in src/prefs/prefs_page_divers.c
Auto-merging src/gsb_scheduler.c
CONFLICT (content): Merge conflict in src/gsb_scheduler.c
Auto-merging po/zh_CN.po
[...]_

Is there a convenient way to mark all my local files to be replaced by grisbi/master's ones?

**Comment by LudovicRousseau** [comment]
You can skip a commit in case of conflict during ``git rebase``. Use ``git rebase --skip``.

I guess this big commit https://github.com/xfred81/grisbi/commit/769a08fad2413c9f258cce914a9cfd57e68c10f7 will be problematic during the ``git rebase``. Is should be safe to skip it.

Another way to recover your patches from the branch grisbi_patch_for_win_targets_ok is to use ``git cherry-pick``.  Create a new branch from master:
```
git checkout master
git checkout -b grisbi_patch_for_windows
```
For each patch you want to keep from https://github.com/xfred81/grisbi/commits/grisbi_patch_for_win_targets_ok you apply it on the current branch grisbi_patch_for_windows using ``git cherry-pick commitID``. Like ``git cherry-pick 4a175c0eddd1256e1a2b0f14b0dec2b35db3552e``.

**Comment by xfred81** [comment]
First issue is on "Applying: Mainstream uodates (#6)", which affects src/categories_onglet.c and src/prefs/prefs_page_divers.c, for example. If I skip it, I do not want changes for official master to be lost; the "skip" option sounds inappropriate thus. 

I tried the -b stuff to look for second solution, but getting each single commits from a web list and applying it on branch manually seems unsatisfactory to me (quite long and, moreover, error prone).

Basically, idea was not to skip identified patches existing on official master (which it seems you suggest), but instead ignoring/replacing changes from my master by those on yours so that conflicts to be removed.

Sorry again for all of this... I'll try to get more knowledge on git on my own.

**Comment by LudovicRousseau** [comment]
NO change from master will ever be lost. The idea is to skip patches that conflict and you are not the author of.

Surprisingly I was able to rebase your branch grisbi_patch_for_win_targets_ok above master with no error or issue.
You can review my branch at https://github.com/LudovicRousseau/grisbi/commits/grisbi_patch_for_win_targets_ok

**Comment by xfred81** [comment]
I tried to apply your suggestion and rebase my master.  Let me know if I'm on the right direction ;)
FYI, command I was looking for to ignore changes (eg. in my local *.po files) was git checkout HEAD <files>

**Comment by LudovicRousseau** [comment]
Your grisbi_patch_for_win_targets_ok branch is in a better shape.
You should not have "Merge" commit in your branch. But only patches above master so they are easy to review.

Another very nice feature of git is to be able to "squash" 2 (or more) patches into 1.
So you can integrate all your test/debug commits and minor changes into 1 change that could be "fix AppVeyor to..."
See http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html for example.

**Comment by xfred81** [comment]
Oh it seems my tries on MSYS2 here are reported here... well, again, I tried to rebase my main branch vs grisbi/master's. And also to integrate my own patches from grisbi_patch_for_win_targets_ok... Hopefully, at least, I won't have broken anything on your side...

**Comment by LudovicRousseau** [comment]
The good news is that you can't break anything on "our" side unless we merge something from you.

The commit "Mainstream uodates (#6)" contains a **lot** of changes and I suspect only some changes are needed. For example this commit changes ``Xcode/config.h`` and I don't think it is intended.

I propose you to:
1. rename grisbi to grisbi.wip (wip for work in progress)
2. start with a clean new clone of grisbi
3. create a branch msys2 to put your changes
3. compare grisbi and grisbi.wip
4. for each difference you want to reuse create a commit in the new grisbi
5. goto step 3

**Comment by xfred81** [comment]
No, changes on Xcode/config.h were not intended to be included.
1. I've renamed grisbi to grisbi.wip in github.com
2. do you mean a new clone out of github? It seems impossible to fork to times from the same project.

I've tried to make an "hard reset" of my master according to grisbi/grisbi's master, but I got tons of "changes" (eg. famous Xcode/config.h) that appeared to be merged, then, vs my former grisbi_patch_for_win_targets_ok branch.

**Comment by LudovicRousseau** [comment]
The rename was for the folder on your disk, not the project name on github.com. Rename it back to grisbi on github.com

I will try an new method. The idea is to have 3 folders on your disk
1. a clean clone of grisbi/master
2. your code to compare with the clean version
3. a new (clean) tree to apply your patches

What you do:
- go into the grisbi directory
- ``git checkout master``
- ``cd ..``
- ``cp -a grisbi grisbi.old``
- ``cp -a grisbi grisbi.new``
- ``cd grisbi.old``
- ``git checkout grisbi_patch_for_win_targets_ok``
- ``cd ../grisbi.new``
- ``git checkout -b new-branch``
- ``cd ..``
- ``diff -ru grisbi.old grisbi``
- for each changes between ``grisbi.old`` and ``grisbi`` directory and that you want to keep create a commit in the branch ``new-branch`` of ``grisbi.new``.

Create small commits instead of a big one as they are easier to manage.

**Comment by LudovicRousseau** [comment]
I note that xfred81/master is not the same as grisbi/master. You master branch contains patches that are **not** in grisbi/master.
In the commands above, before you can compare grisbi.old and grisbi you need to do:
``cd grisbi ; git checkout upstream/master`` so that the grisbi directory contains the same code as upstream/master.

**Comment by xfred81** [comment]
Ok. Finally tried this; 

```
$ git checkout upstream/master
Note: checking out 'upstream/master'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout
_[...]_
```

I suppose that's part of my problem. But I don't see why I'm in detached HEAD state. It looks like it'd be much more easier to start all over. The diffs are really huge.

[diff.ru.gz](https://github.com/grisbi/grisbi/files/1782806/diff.ru.gz)

And most of them are not related to my work, but probably that I've been in "detached HEAD" state for a too long time :( Also, many intermediate commits from grisbi/master that I've pulled in my master are not recognized.

Hm... I really wonder if I'd not gain some time by restarting the whole story from scratch (?)

**Comment by LudovicRousseau** [comment]
Sometimes it is faster to start again with a clean repository.

**Comment by xfred81** [comment]
HOoah. I finally come up with a new, clean git repository on my side (well, hopefully).
I've just rebuilt 32 & 64-bit Windows versions, but grisbi.exe just halt when started (either when I use my local build or the AppVeyor's).

Did you experience such behvaviour on your side? it looks like missing dependencies (eg. AFAIR former installer was much larger - don't we miss a data file or such?), but I don't have much time to continue tests right now.

Again, thanks for your patience/support guys.

**Comment by xfred81** [comment]
I don't know if this helps, but level 5 debuging returns:

```
> [...]
> Thu Mar 22 23:42:11 2018, 0,390000 : Debug - ../../grisbi.xf/src/import.c:1329:gsb_import_register_ImportFormat - Adding 'Gnucash' as an import format
> Thu Mar 22 23:42:11 2018, 0,390000 : Debug - ../../grisbi.xf/src/grisbi_app.c:409:grisbi_app_init_recent_files_menu
> Thu Mar 22 23:42:11 2018, 0,390000 : Debug - ../../grisbi.xf/src/grisbi_app.c:974:grisbi_app_activate
> Thu Mar 22 23:42:11 2018, 0,406000 : Debug - ../../grisbi.xf/src/grisbi_win.c:622:grisbi_win_no_file_page_new
```

and program halts just after last line - or, well, at least, no more information is print()ed.

**Comment by xfred81** [comment]
gdb tells program halts on "singleton" initialisation:

```
GrisbiSettings *grisbi_settings_get (void)
{
    if (!singleton)
    {
        singleton = GRISBI_SETTINGS (g_object_new (GRISBI_TYPE_SETTINGS, NULL));
    }
    else ...

```

**Comment by LudovicRousseau** [comment]
Welcome back @xfred81 
I do not use Windows so I can't really help.

Maybe you can compare the previous (working) installer with the new one to see what files are missing?

**Comment by LudovicRousseau** [comment]
The best would be to continue the discussion on the grisbi devel mailing list.
http://listes.grisbi.org/mailman/listinfo/devel

**Comment by xfred81** [comment]
Any feature to anonymize mail (eg. using githut.com)?
'Cause I tried the rebasing stuff, now my repository *seems* OK (branch msys2 & master).
Also, I've produced Windows binaries with latest source from upstream (grisbi/grisbi).

That's just for information. OK, I promise I'll use mailing list (well, if any chance of hiding my mail).

Yet, just for fun, I've just re-launched a "git rebase -Xours upstream/master" and it now founds 100+ or so different commits to be applied (including my own very latest commits)... Rebasing is really obscure, and despite many times spent on reading documentation, I never get results expected (and attended, according to my understanding of "theory").

**No need to reply**. I'm more than aware that I probably the only guy out there facing such issues (BTW, IMO, skipping patches during rebase is *not* a good thing).

**Comment by LudovicRousseau** [comment]
Ton email n'est pas celui indiqué dans https://github.com/xfred81/grisbi/commit/755c85b5cb3976df92e6b5c47caf930e60a6476e (entre autres commit) ?

Je pense que tes problèmes avec "git rebase" viennent du fait que ta branche a des conflits avec le dépôt upstream.
Appliquer des patches du dépôts upstream et faire des merges avec upstream et une très mauvaise idée si tu veux que "git rebase" se passe bien.

Tes problèmes ne vont que s'empirer avec les modifications.

**Comment by xfred81** [comment]
Indeed. The more I (try to) contribute the worst it gets.
My points is pulling regularly from upstream should avoid conflicts.
